File: workers/research/src/black_book_research/resolution/resolver.py
"""Deterministic resolution scoring that queues ambiguity and never merges silently."""

from __future__ import annotations

from dataclasses import replace

from .models import (
    DuplicateReviewQueueItem,
    EntityProfile,
    Jurisdiction,
    MatchFactor,
    RankedEntityMatch,
    ResolutionCandidate,
    ResolutionDecision,
    ResolutionResult,
)
from .normalization import name_similarity, normalize_alias, parse_address

PROPOSE_THRESHOLD = 0.86
REVIEW_THRESHOLD = 0.55
AMBIGUITY_MARGIN = 0.12
ORGANIZATION_SUFFIXES = frozenset(
    {
        "association",
        "company",
        "corporation",
        "inc",
        "incorporated",
        "llc",
        "organization",
        "society",
    }
)


def _active(year: int, valid_from: int | None, valid_to: int | None) -> bool:
    return (valid_from is None or year >= valid_from) and (
        valid_to is None or year <= valid_to
    )


def _organization_name(value: str) -> str:
    tokens = normalize_alias(value).split()
    while len(tokens) > 1 and tokens[-1] in ORGANIZATION_SUFFIXES:
        tokens.pop()
    return " ".join(tokens)


def _score_name(candidate: ResolutionCandidate, entity: EntityProfile) -> MatchFactor:
    dated_names = entity.aliases + entity.school_names
    canonical_names = [entity.display_name]
    canonical_names.extend(
        item.name
        for item in dated_names
        if candidate.year is None
        or _active(candidate.year, item.valid_from, item.valid_to)
    )
    best = 0.0
    pair = (candidate.name, entity.display_name)
    for candidate_name in (candidate.name, *candidate.aliases):
        for canonical_name in canonical_names:
            direct = name_similarity(candidate_name, canonical_name)
            organization = (
                name_similarity(
                    _organization_name(candidate_name),
                    _organization_name(canonical_name),
                )
                if candidate.kind == "organization" or entity.kind == "organization"
                else 0.0
            )
            score = max(direct, organization)
            if score > best:
                best, pair = score, (candidate_name, canonical_name)
    method = "exact normalized" if best == 1 else "fuzzy"
    return MatchFactor(
        "name",
        best * 0.55,
        f'{method} name match "{pair[0]}" ↔ "{pair[1]}" ({best:.3f})',
    )


def _score_kind(candidate: ResolutionCandidate, entity: EntityProfile) -> MatchFactor:
    if candidate.kind is None:
        return MatchFactor("kind", 0.08, "candidate kind unspecified")
    if candidate.kind == entity.kind:
        return MatchFactor("kind", 0.15, f"entity kind matches {candidate.kind}")
    return MatchFactor(
        "kind", -0.35, f"entity kind conflict: {candidate.kind} vs {entity.kind}"
    )


def _score_identifier(
    candidate: ResolutionCandidate, entity: EntityProfile
) -> MatchFactor:
    canonical = {
        (normalize_alias(system), normalize_alias(value))
        for system, value in entity.identifiers
    }
    match = next(
        (
            system
            for system, value in candidate.identifiers
            if (normalize_alias(system), normalize_alias(value)) in canonical
        ),
        None,
    )
    if match:
        return MatchFactor("identifier", 0.1, f"exact identifier match for {match}")
    rationale = (
        "no identifier match"
        if candidate.identifiers
        else "no candidate identifier supplied"
    )
    return MatchFactor("identifier", 0.0, rationale)


def _score_geography(
    candidate: ResolutionCandidate,
    entity: EntityProfile,
    jurisdictions: tuple[Jurisdiction, ...],
) -> MatchFactor:
    parsed = parse_address(candidate.address) if candidate.address else {}
    hints = [
        *candidate.geographic_hints,
        *([parsed["city"]] if "city" in parsed else []),
        *([parsed["state"], f"US-{parsed['state']}"] if "state" in parsed else []),
    ]
    normalized_hints = [
        normalize_alias(hint) for hint in hints if normalize_alias(hint)
    ]
    if not normalized_hints:
        return MatchFactor("geography", 0.05, "no geographic evidence supplied")
    valid_jurisdictions = {
        jurisdiction.id: jurisdiction.name
        for jurisdiction in jurisdictions
        if candidate.year is None
        or _active(candidate.year, jurisdiction.valid_from, jurisdiction.valid_to)
    }
    values: list[str] = []
    for location in entity.locations:
        if candidate.year is not None and not _active(
            candidate.year, location.valid_from, location.valid_to
        ):
            continue
        values.append(normalize_alias(location.label))
        values.extend(
            normalize_alias(valid_jurisdictions[jurisdiction_id])
            for jurisdiction_id in location.jurisdiction_ids
            if jurisdiction_id in valid_jurisdictions
        )
    matches = any(
        hint in value or value in hint
        for hint in normalized_hints
        for value in values
        if value
    )
    return MatchFactor(
        "geography",
        0.12 if matches else -0.08,
        "candidate geography matches a valid entity location or jurisdiction"
        if matches
        else "candidate geography conflicts with known entity locations",
    )


def _score_temporal(
    candidate: ResolutionCandidate, entity: EntityProfile
) -> MatchFactor:
    if candidate.year is None:
        return MatchFactor("temporal", 0.05, "candidate year unspecified")
    if (entity.birth_year is not None and candidate.year < entity.birth_year) or (
        entity.death_year is not None and candidate.year > entity.death_year
    ):
        return MatchFactor(
            "temporal", -0.4, "candidate year falls outside person lifespan"
        )
    valid_locations = [
        location
        for location in entity.locations
        if _active(candidate.year, location.valid_from, location.valid_to)
    ]
    if entity.locations and not valid_locations:
        return MatchFactor(
            "temporal",
            -0.2,
            "no historical/current location is valid in candidate year",
        )
    role = (
        "historical"
        if any(location.role == "historical" for location in valid_locations)
        else "current"
    )
    return MatchFactor(
        "temporal",
        0.08,
        f"{role} evidence is consistent with candidate year {candidate.year}",
    )


def score_entity_match(
    candidate: ResolutionCandidate,
    entity: EntityProfile,
    jurisdictions: tuple[Jurisdiction, ...] = (),
) -> RankedEntityMatch:
    factors = (
        _score_name(candidate, entity),
        _score_kind(candidate, entity),
        _score_identifier(candidate, entity),
        _score_geography(candidate, entity, jurisdictions),
        _score_temporal(candidate, entity),
    )
    confidence = round(max(0.0, min(1.0, sum(factor.score for factor in factors))), 4)
    return RankedEntityMatch(entity.id, confidence, factors)


def resolve_entity_candidate(
    candidate: ResolutionCandidate,
    entities: tuple[EntityProfile, ...],
    jurisdictions: tuple[Jurisdiction, ...] = (),
) -> ResolutionResult:
    ranked = tuple(
        sorted(
            (
                score_entity_match(candidate, entity, jurisdictions)
                for entity in entities
            ),
            key=lambda match: (-match.confidence, match.entity_id),
        )
    )
    if not ranked or ranked[0].confidence < REVIEW_THRESHOLD:
        return ResolutionResult(
            candidate.id,
            "no_match",
            None,
            ranked,
            ("No canonical entity reached the review confidence threshold.",),
        )
    ambiguous = (
        len(ranked) > 1
        and ranked[0].confidence - ranked[1].confidence < AMBIGUITY_MARGIN
    )
    if ranked[0].confidence < PROPOSE_THRESHOLD or ambiguous:
        rationale = (
            "Top matches are within the ambiguity margin; no entity was selected."
            if ambiguous
            else "Best match requires human review; no entity was selected."
        )
        return ResolutionResult(
            candidate.id, "review_required", None, ranked, (rationale,)
        )
    return ResolutionResult(
        candidate.id,
        "proposed_match",
        ranked[0].entity_id,
        ranked,
        tuple(factor.rationale for factor in ranked[0].factors),
    )


def create_duplicate_review_item(
    candidate: ResolutionCandidate,
    result: ResolutionResult,
    created_at: str,
) -> DuplicateReviewQueueItem:
    if result.outcome != "review_required":
        raise ValueError(f"Candidate {candidate.id} is not review-required")
    ambiguous = len(result.ranked_matches) > 1 and (
        result.ranked_matches[0].confidence - result.ranked_matches[1].confidence
        < AMBIGUITY_MARGIN
    )
    return DuplicateReviewQueueItem(
        id=f"resolution-review:{candidate.id}",
        candidate_id=candidate.id,
        candidate_name=candidate.name,
        reason="ambiguous_match" if ambiguous else "low_confidence_match",
        proposed_entity_ids=tuple(
            match.entity_id for match in result.ranked_matches[:3]
        ),
        ranked_matches=result.ranked_matches,
        source_reference_ids=candidate.source_reference_ids,
        created_at=created_at,
    )


def apply_resolution_decision(
    decision: ResolutionDecision, applied_at: str
) -> ResolutionDecision:
    if decision.status != "proposed":
        raise ValueError(f"Resolution decision {decision.id} is not proposed")
    return replace(decision, status="applied", applied_at=applied_at)


def reverse_resolution_decision(
    decision: ResolutionDecision,
    *,
    reversed_at: str,
    reversed_by: str,
    reason: str,
) -> ResolutionDecision:
    if decision.status != "applied":
        raise ValueError(f"Resolution decision {decision.id} is not applied")
    return replace(
        decision,
        status="reversed",
        reversed_at=reversed_at,
        reversed_by=reversed_by,
        reverse_reason=reason,
    )

File: packages/domain/src/resolution/resolver.ts
/**
 * Scores entity candidates across names, identifiers, time, and geography without merging records.
 */
import type { DiscoveryCandidateRecord } from '../discovery/types.js';
import type { CanonicalEntity, EntityAlias } from '../entity.js';
import { isEntityKind } from '../entity-kinds.js';
import type { EntityLocation, Jurisdiction } from '../geography/location.js';
import { isTrustedIdentifierNamespace } from '../naming.js';
import {
  nameSimilarity,
  normalizeAlias,
  normalizeOrganizationName,
  parseAddress,
} from './normalization.js';
import type {
  DuplicateReviewQueueItem,
  MatchFactor,
  RankedEntityMatch,
  ResolutionCandidate,
  ResolutionContext,
  ResolutionDecision,
  ResolutionProfile,
  ResolutionResult,
} from './types.js';

const PROPOSE_THRESHOLD = 0.86;
const REVIEW_THRESHOLD = 0.55;
const AMBIGUITY_MARGIN = 0.12;

/**
 * black-book-8bck (identifier-dominant resolution): an exact match on a TRUSTED-namespace
 * identifier (Wikidata QID, LoC, VIAF, NPS, NRHP, NCES, etc. — see `../naming.js`) must clearly
 * outrank name similarity, whose maximum contribution is 0.55 (see `nameFactor` below). 0.65
 * dominates that ceiling on its own. An exact match on an UNTRUSTED/internal namespace keeps the
 * prior, smaller weight (0.1) — unchanged from before this bead, since an internal accession
 * number is not unambiguous evidence the way an external authority-control id is.
 */
const EXACT_TRUSTED_IDENTIFIER_SCORE = 0.65;
const EXACT_UNTRUSTED_IDENTIFIER_SCORE = 0.1;

function yearFromDate(value: string | undefined | null): number | undefined {
  if (!value) return undefined;
  const match = value.match(/^\d{4}/);
  return match ? Number(match[0]) : undefined;
}

function activeInYear(
  year: number,
  value: { readonly validFrom?: string; readonly validTo?: string | null },
): boolean {
  const from = yearFromDate(value.validFrom);
  const to = yearFromDate(value.validTo);
  return (from === undefined || year >= from) && (to === undefined || year <= to);
}

function entityNames(entity: CanonicalEntity, year?: number): readonly string[] {
  const aliases: readonly EntityAlias[] = entity.aliases ?? [];
  const names = [
    entity.displayName,
    ...aliases
      .filter((alias) => year === undefined || activeInYear(year, alias))
      .map((alias) => alias.value),
    ...(entity.school?.names ?? [])
      .filter((name) => year === undefined || activeInYear(year, name))
      .map((name) => name.name),
  ];
  return [...new Set(names)];
}

function nameFactor(candidate: ResolutionCandidate, entity: CanonicalEntity): MatchFactor {
  const candidateNames = [candidate.name, ...(candidate.aliases ?? [])];
  const canonicalNames = entityNames(entity, candidate.year);
  let best = 0;
  let matchedCandidate = candidate.name;
  let matchedCanonical = entity.displayName;
  for (const candidateName of candidateNames) {
    for (const canonicalName of canonicalNames) {
      const direct = nameSimilarity(candidateName, canonicalName);
      const organization =
        candidate.kind === 'organization' || entity.kind === 'organization'
          ? nameSimilarity(
              normalizeOrganizationName(candidateName),
              normalizeOrganizationName(canonicalName),
            )
          : 0;
      const score = Math.max(direct, organization);
      if (score > best) {
        best = score;
        matchedCandidate = candidateName;
        matchedCanonical = canonicalName;
      }
    }
  }
  const method = best === 1 ? 'exact normalized' : 'fuzzy';
  return {
    factor: 'name',
    score: best * 0.55,
    rationale: `${method} name match "${matchedCandidate}" ↔ "${matchedCanonical}" (${best.toFixed(3)})`,
  };
}

function identifierFactor(candidate: ResolutionCandidate, entity: CanonicalEntity): MatchFactor {
  const candidateIdentifiers = Object.entries(candidate.identifiers ?? {});
  const match = candidateIdentifiers.find(([system, value]) =>
    entity.identifiers?.some(
      (identifier) =>
        normalizeAlias(identifier.system) === normalizeAlias(system) &&
        normalizeAlias(identifier.value) === normalizeAlias(value),
    ),
  );
  if (!match) {
    return {
      factor: 'identifier',
      score: 0,
      rationale: candidateIdentifiers.length ? 'no identifier match' : 'no candidate identifier supplied',
    };
  }
  const trusted = isTrustedIdentifierNamespace(match[0]);
  return {
    factor: 'identifier',
    score: trusted ? EXACT_TRUSTED_IDENTIFIER_SCORE : EXACT_UNTRUSTED_IDENTIFIER_SCORE,
    rationale: trusted
      ? `exact identifier match on trusted namespace ${match[0]} outranks name similarity`
      : `exact identifier match for ${match[0]}`,
  };
}

function kindFactor(candidate: ResolutionCandidate, entity: CanonicalEntity): MatchFactor {
  if (!candidate.kind) {
    return { factor: 'kind', score: 0.08, rationale: 'candidate kind unspecified' };
  }
  return candidate.kind === entity.kind
    ? { factor: 'kind', score: 0.15, rationale: `entity kind matches ${candidate.kind}` }
    : {
        factor: 'kind',
        score: -0.35,
        rationale: `entity kind conflict: ${candidate.kind} vs ${entity.kind}`,
      };
}

function locationText(location: EntityLocation): string {
  return normalizeAlias(
    [location.label, location.modernZip?.zip, ...(location.jurisdictionIds ?? [])]
      .filter(Boolean)
      .join(' '),
  );
}

function relevantJurisdictions(
  jurisdictions: readonly Jurisdiction[],
  year: number | undefined,
): readonly Jurisdiction[] {
  return jurisdictions.filter(
    (jurisdiction) => year === undefined || activeInYear(year, jurisdiction),
  );
}

function geographyFactor(
  candidate: ResolutionCandidate,
  locations: readonly EntityLocation[],
  context: ResolutionContext,
): MatchFactor {
  const parsed = candidate.address ? parseAddress(candidate.address) : undefined;
  const hints = [
    ...(candidate.geographicHints ?? []),
    ...(parsed?.city ? [parsed.city] : []),
    ...(parsed?.state ? [parsed.state, `US-${parsed.state}`] : []),
  ]
    .map(normalizeAlias)
    .filter(Boolean);
  if (!hints.length) {
    return { factor: 'geography', score: 0.05, rationale: 'no geographic evidence supplied' };
  }
  const jurisdictions = relevantJurisdictions(context.jurisdictions ?? [], candidate.year);
  const jurisdictionNames = new Map(
    jurisdictions.flatMap((jurisdiction) => [
      [normalizeAlias(jurisdiction.id), normalizeAlias(jurisdiction.name)],
      [normalizeAlias(jurisdiction.name), normalizeAlias(jurisdiction.name)],
    ]),
  );
  const validLocations = locations.filter(
    (location) => candidate.year === undefined || activeInYear(candidate.year, location),
  );
  const locationValues = validLocations.flatMap((location) => {
    const values = [locationText(location)];
    for (const id of location.jurisdictionIds ?? []) {
      const name = jurisdictionNames.get(normalizeAlias(id));
      if (name) values.push(name);
    }
    return values;
  });
  const matches = hints.some((hint) =>
    locationValues.some((value) => value.includes(hint) || hint.includes(value)),
  );
  return {
    factor: 'geography',
    score: matches ? 0.12 : -0.08,
    rationale: matches
      ? 'candidate geography matches a valid entity location or jurisdiction'
      : 'candidate geography conflicts with known entity locations',
  };
}

function temporalFactor(
  candidate: ResolutionCandidate,
  entity: CanonicalEntity,
  locations: readonly EntityLocation[],
): MatchFactor {
  if (candidate.year === undefined) {
    return { factor: 'temporal', score: 0.05, rationale: 'candidate year unspecified' };
  }
  const impossiblePerson =
    entity.person &&
    ((entity.person.birthYear !== undefined &&
      entity.person.birthYear !== null &&
      candidate.year < entity.person.birthYear) ||
      (entity.person.deathYear !== undefined &&
        entity.person.deathYear !== null &&
        candidate.year > entity.person.deathYear));
  if (impossiblePerson) {
    return {
      factor: 'temporal',
      score: -0.4,
      rationale: 'candidate year falls outside person lifespan',
    };
  }
  const validLocations = locations.filter((location) => activeInYear(candidate.year!, location));
  if (locations.length && !validLocations.length) {
    return {
      factor: 'temporal',
      score: -0.2,
      rationale: 'no historical/current location is valid in candidate year',
    };
  }
  const role = validLocations.some((location) => location.role === 'historical')
    ? 'historical'
    : validLocations.some((location) => location.role === 'current')
      ? 'current'
      : 'dated';
  return {
    factor: 'temporal',
    score: 0.08,
    rationale: `${role} evidence is consistent with candidate year ${candidate.year}`,
  };
}

export function scoreEntityMatch(
  candidate: ResolutionCandidate,
  profile: ResolutionProfile,
  context: ResolutionContext = {},
): RankedEntityMatch {
  const factors = [
    nameFactor(candidate, profile.entity),
    kindFactor(candidate, profile.entity),
    identifierFactor(candidate, profile.entity),
    geographyFactor(candidate, profile.locations ?? [], context),
    temporalFactor(candidate, profile.entity, profile.locations ?? []),
  ];
  const confidence = Math.max(
    0,
    Math.min(
      1,
      factors.reduce((sum, factor) => sum + factor.score, 0),
    ),
  );
  return { entityId: profile.entity.id, confidence: Number(confidence.toFixed(4)), factors };
}

export function resolveEntityCandidate(
  candidate: ResolutionCandidate,
  profiles: readonly ResolutionProfile[],
  context: ResolutionContext = {},
): ResolutionResult {
  const rankedMatches = profiles
    .map((profile) => scoreEntityMatch(candidate, profile, context))
    .sort(
      (left, right) =>
        right.confidence - left.confidence || left.entityId.localeCompare(right.entityId),
    );
  const best = rankedMatches[0];
  const runnerUp = rankedMatches[1];
  if (!best || best.confidence < REVIEW_THRESHOLD) {
    return {
      candidateId: candidate.id,
      outcome: 'no_match',
      rankedMatches,
      rationale: ['No canonical entity reached the review confidence threshold.'],
    };
  }
  const ambiguous =
    runnerUp !== undefined && best.confidence - runnerUp.confidence < AMBIGUITY_MARGIN;
  if (best.confidence < PROPOSE_THRESHOLD || ambiguous) {
    return {
      candidateId: candidate.id,
      outcome: 'review_required',
      rankedMatches,
      rationale: [
        ambiguous
          ? 'Top matches are within the ambiguity margin; no entity was selected.'
          : 'Best match requires human review; no entity was selected.',
      ],
    };
  }
  return {
    candidateId: candidate.id,
    outcome: 'proposed_match',
    selectedEntityId: best.entityId,
    rankedMatches,
    rationale: best.factors.map((factor) => factor.rationale),
  };
}

function payloadStrings(
  payload: Readonly<Record<string, unknown>> | undefined,
  key: string,
): readonly string[] {
  const value = payload?.[key];
  if (typeof value === 'string') return [value];
  if (Array.isArray(value)) return value.filter((item): item is string => typeof item === 'string');
  return [];
}

export function resolutionCandidateFromDiscovery(
  candidate: DiscoveryCandidateRecord,
): ResolutionCandidate {
  const payload = candidate.adapterRecord.payload;
  const yearValue = payload?.year;
  const identifiersValue = payload?.identifiers;
  const identifiers =
    identifiersValue && typeof identifiersValue === 'object' && !Array.isArray(identifiersValue)
      ? Object.fromEntries(
          Object.entries(identifiersValue).filter(
            (entry): entry is [string, string] => typeof entry[1] === 'string',
          ),
        )
      : undefined;
  const address = payloadStrings(payload, 'address')[0];
  const year = typeof yearValue === 'number' ? yearValue : undefined;
  const kindValue = payloadStrings(payload, 'kind')[0];
  return {
    id: candidate.id,
    name: payloadStrings(payload, 'name')[0] ?? candidate.adapterRecord.title ?? '',
    ...(kindValue !== undefined && isEntityKind(kindValue) ? { kind: kindValue } : {}),
    aliases: payloadStrings(payload, 'aliases'),
    ...(address !== undefined ? { address } : {}),
    ...(year !== undefined ? { year } : {}),
    geographicHints: candidate.geographicHints.map((hint) => hint.text),
    ...(identifiers !== undefined ? { identifiers } : {}),
    sourceReferenceIds: candidate.identity.sourceReferences.map(
      (reference) => `${reference.sourceId}:${reference.stableIdentifier}`,
    ),
  };
}

export function createDuplicateReviewQueueItem(
  candidate: ResolutionCandidate,
  result: ResolutionResult,
  createdAt: string,
): DuplicateReviewQueueItem {
  if (result.outcome !== 'review_required') {
    throw new Error(`Candidate ${candidate.id} is not review-required`);
  }
  return {
    id: `resolution-review:${candidate.id}`,
    candidateId: candidate.id,
    candidateName: candidate.name,
    status: 'pending',
    reason:
      result.rankedMatches[1] &&
      result.rankedMatches[0]!.confidence - result.rankedMatches[1].confidence < AMBIGUITY_MARGIN
        ? 'ambiguous_match'
        : 'low_confidence_match',
    proposedEntityIds: result.rankedMatches.slice(0, 3).map((match) => match.entityId),
    rankedMatches: result.rankedMatches,
    sourceReferenceIds: candidate.sourceReferenceIds,
    createdAt,
  };
}

export function applyResolutionDecision(
  decision: ResolutionDecision,
  appliedAt: string,
): ResolutionDecision {
  if (decision.status !== 'proposed') {
    throw new Error(`Resolution decision ${decision.id} is not proposed`);
  }
  return { ...decision, status: 'applied', appliedAt };
}

export function reverseResolutionDecision(
  decision: ResolutionDecision,
  input: { readonly reversedAt: string; readonly reversedBy: string; readonly reason: string },
): ResolutionDecision {
  if (decision.status !== 'applied') {
    throw new Error(`Resolution decision ${decision.id} is not applied`);
  }
  return {
    ...decision,
    status: 'reversed',
    reversedAt: input.reversedAt,
    reversedBy: input.reversedBy,
    reverseReason: input.reason,
  };
}
